rfc-160-using-es6-modules-in-govuk_publishing_components.md

```diff
@@ -0,0 +1,428 @@
+
+##### 2. Converting the Javascript to ES6
+
+In govuk_publishing_components we use a ['module'-like implementation](https://docs.publishing.service.gov.uk/repos/govuk_publishing_components/javascript-modules.html) in which the Javascript of each component is self-contained in a function. This function is then assigned to the GOVUK window object and then the components are initialised using an "initAll" method which iterates through the components and runs each initialisation method.
+
+As this 'module'-like pattern has already been implemented, the actual conversion to ES6 is straight forward. It would consist of wrapping the existing code in a `Class` and removing the `prototype` syntax.
+
+```
+(function (Modules) {
+  function ModuleName ($module) {
+    this.$module = $module;
+  }
+
+  ModuleName.prototype.utilityFunction = function () {}
+
+  ModuleName.prototype.init = function () {}
+})(window.GOVUK.Modules)
+```
+to 
+
+```
+Class ModuleName {
+  constructor($module) {
+    this.$module = $module;
+  }
+
+  utilityFunction() {}
+
+  init() {
+    this.utilityFunction();
+    // rest of the init code...
+  }
+}
+
+export default ModuleName;
+``` 
+
+In govuk_publishing_components, Javascript is used to add extra functionality (usually analytics) to components implemented in govuk-frontend. Currently the approach is to first initialise the govuk-frontend Javascript module of a component and then initialise the govuk_publishing_components Javascript module to add - for instance - extra analytics when a user clicks specific areas of the component. 
+
+```
+window.GOVUK.Component.init(exampleElement);
+window.GOVUK.GemComponent.init(exampleElement);
+```
+
+With jsbundling we can simply this initialisation by using composition: 
+
+```
+import { Component } from 'govuk-frontend';
+
+Class GemComponent {
+  constructor($module) {
+    this.govukFrontendComponent = new Component($module); // assign a variable to an instance of the component
+    this.classNameOfElement = "class-name-of-element";
+    // specific constructor code for the gem component
+  }
+
+  init() {
+    this.govukFrontendComponent.init();  // initialise the instance of the component
+    // specific init code for the gem component
+  }
+}
+
+export default GemComponent;
+```
+
+```
+window.GOVUK.GemComponent.init(exampleElement);
```

Comment:
Thanks for the detailed talk this morning @patrickpatrickpatrick 😊

As this is similar to our `window.GOVUKFrontend` global I'll add some of our findings here

We realised a while back that it was the Rollup UMD bundle that added globals for:

```mjs
window.GOVUKFrontend.Accordion
window.GOVUKFrontend.CharacterCount
window.GOVUKFrontend.Checkboxes
window.GOVUKFrontend.Radios
// etc
```

Whilst we're still going to publish UMD bundles in `govuk-frontend@5` we'd prefer if teams used ES6 module `import`/`export` syntax too, rather than put UMD bundles in a `<script type="module">` wrapper

### Before 
Rollup UMD bundles automatically add `window` globals

```html
<script src="/govuk-frontend.min.js"></script>
<script>
  window.GOVUKFrontend.initAll()
</script>
```

### After
Rollup ES bundles use ES6 `import`/`export` syntax

```html
<script type="module">
  import { initAll } from '/govuk-frontend.min.js'
  initAll()
</script>
```

Might be a while before you can move away from `window` globals

But by transpiling ES6 to ES5 (using [Bublé](https://github.com/bublejs/buble#readme)) you'd lose access to the good stuff™ like exports 😊

Reply:
I should add that ES module downloads will be deferred by default

So we'd recommend they still be "preloaded" with a script tag before import

```html
<script type="module" src="/govuk-frontend.min.js"></script>
```

Or using [`modulepreload` hints where supported](https://caniuse.com/link-rel-modulepreload)

```html
<link rel="modulepreload" href="/govuk-frontend.min.js">
```

Reply:
With regards to transpiling at all, this is because frontend rendering applications don't have support for using ES6 modules as they are using sprockets and asset pipeline. If we were to make[ the significant change](https://github.com/alphagov/govuk-rfcs/blob/aa16e2b5c0f32784dc83117dc0c90d1d4fbe73c8/rfc-160-using-es6-modules-in-govuk_publishing_components.md#significant-change-move-all-javascript-to-a-separate-npm-module) (moving the Javascript outside of the gem entirely) then this would require all the frontend rendering applications to be updated to using bundlers so we could actually make use of modules and actual ES6 features in Rails applications. My thinking of this RFC was the least disruptive way to our frontend applications that would make govuk_publishing_components compatible with ES6.

Reply:
Definitely. Mainly commenting to show what's coming in future

>Might be a while before you can move away from `window` globals

But hopefully can avoid transpiling to ES5 anything inside ES2015 `<script type="module">`

Reply:
That would make sense actually, since it wouldn't run without polyfill anyway we could just strip out the transpiling step for components. Whether to include transpiling at all depends on if we just leave analytics as they are (which I think we could, unless we want to transpire them but that could start to introducing polyfill for them which I assume would want to be avoided). 

This would mean rethinking minification as Uglify is only compatible with ES5. There are alternatives out there (https://github.com/ahorek/terser-ruby) we could use so it's not a big issue. Would just mean integration would be more complex as we'd have to update the frontend rendering applications simultaneously with the changes to govuk_publishing_components.